File: src/main/java/net/dontdrinkandroot/gitki/wicket/form/UserEditForm.java
package net.dontdrinkandroot.gitki.wicket.form;

import net.dontdrinkandroot.gitki.model.Role;
import net.dontdrinkandroot.gitki.model.User;
import net.dontdrinkandroot.gitki.service.user.UserService;
import net.dontdrinkandroot.gitki.wicket.GitkiWebSession;
import net.dontdrinkandroot.gitki.wicket.choicerenderer.IsoLanguageChoiceRenderer;
import net.dontdrinkandroot.gitki.wicket.choicerenderer.ZoneIdChoiceRenderer;
import net.dontdrinkandroot.gitki.wicket.page.admin.UserListPage;
import net.dontdrinkandroot.wicket.bootstrap.component.button.SubmitLabelButton;
import net.dontdrinkandroot.wicket.bootstrap.component.form.RepeatingAjaxForm;
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputEmail;
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputPassword;
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputText;
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupSelect;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.model.PropertyModel;
import org.apache.wicket.model.StringResourceModel;
import org.apache.wicket.spring.injection.annot.SpringBean;

import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class UserEditForm extends RepeatingAjaxForm<User>
{
    @SpringBean
    private UserService userService;

    private IModel<String> newPasswordModel = new Model<>();

    public UserEditForm(String id, IModel<User> model)
    {
        super(id, model);
    }

    @Override
    protected void populateFormGroups(RepeatingView formGroupView)
    {
        FormGroupInputText formGroupFirstName = new FormGroupInputText(
                formGroupView.newChildId(),
                Model.of("First Name"),
                new PropertyModel<>(this.getModel(), "firstName")
        );
        formGroupFirstName.setRequired(true);
        formGroupFirstName.addDefaultAjaxInputValidation();
        formGroupView.add(formGroupFirstName);

        FormGroupInputText formGroupLastName = new FormGroupInputText(
                formGroupView.newChildId(),
                Model.of("Last Name"),
                new PropertyModel<>(this.getModel(), "lastName")
        );
        formGroupLastName.setRequired(true);
        formGroupLastName.addDefaultAjaxInputValidation();
        formGroupView.add(formGroupLastName);

        FormGroupInputEmail formGroupEmail =
                new FormGroupInputEmail(
                        formGroupView.newChildId(),
                        Model.of("Email"),
                        new PropertyModel<>(this.getModel(), "email")
                );
        formGroupEmail.setRequired(true);
        formGroupEmail.addDefaultAjaxInputValidation();
        formGroupView.add(formGroupEmail);

        FormGroupSelect<String> formGroupLanguage = new FormGroupSelect<>(
                formGroupView.newChildId(),
                Model.of("Language"),
                new PropertyModel<>(this.getModel(), "language"),
                Arrays.asList(Locale.getISOLanguages()),
                new IsoLanguageChoiceRenderer()
        );
        formGroupLanguage.setRequired(true);
        formGroupView.add(formGroupLanguage);

        List<String> availableZoneIds = new ArrayList<>(ZoneId.getAvailableZoneIds());
        availableZoneIds.sort(String::compareTo);
        FormGroupSelect<String> formGroupZoneId =
                new FormGroupSelect<>(
                        formGroupView.newChildId(),
                        new StringResourceModel("gitki.timezone"),
                        new PropertyModel<>(this.getModel(), "zoneId"),
                        availableZoneIds,
                        new ZoneIdChoiceRenderer()
                );
        formGroupZoneId.setRequired(true);
        formGroupView.add(formGroupZoneId);

        FormGroupSelect<Role> formGroupRole =
                new FormGroupSelect<>(
                        formGroupView.newChildId(),
                        Model.of("Role"),
                        new PropertyModel<>(this.getModel(), "role"),
                        Arrays.asList(Role.values())
                );
        formGroupRole.setNullValid(false);
        formGroupRole.setRequired(true);
        formGroupRole.setVisible(GitkiWebSession.get().hasRole(Role.ADMIN));
        formGroupView.add(formGroupRole);

        FormGroupInputPassword formGroupPassword =
                new FormGroupInputPassword(formGroupView.newChildId(), Model.of("Password"), this.newPasswordModel);
        formGroupPassword.setRequired(false);
        formGroupView.add(formGroupPassword);
    }

    @Override
    protected void populateActions(RepeatingView buttonView)
    {
        buttonView.add(new SubmitLabelButton(buttonView.newChildId(), new StringResourceModel("gitki.save")));
    }

    @Override
    protected void onSubmit(AjaxRequestTarget target)
    {
        super.onSubmit(target);
        User user = this.getModelObject();
        user = this.userService.save(user, this.newPasswordModel.getObject());

        /* Update user in session so changes are reflected e.g. for locale*/
        if (GitkiWebSession.get().getUser().equals(user)) {
            GitkiWebSession.get().signIn(user);
        }
    }

    @Override
    protected void onAfterSubmit(AjaxRequestTarget target)
    {
        super.onAfterSubmit(target);
        this.setResponsePage(UserListPage.class);
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/form/UserEditForm.kt
package net.dontdrinkandroot.gitki.wicket.form

import net.dontdrinkandroot.gitki.model.Role
import net.dontdrinkandroot.gitki.model.User
import net.dontdrinkandroot.gitki.service.user.UserService
import net.dontdrinkandroot.gitki.wicket.choicerenderer.IsoLanguageChoiceRenderer
import net.dontdrinkandroot.gitki.wicket.choicerenderer.ZoneIdChoiceRenderer
import net.dontdrinkandroot.gitki.wicket.getCurrentUser
import net.dontdrinkandroot.gitki.wicket.getGitkiSession
import net.dontdrinkandroot.wicket.behavior.CssClassAppender
import net.dontdrinkandroot.wicket.bootstrap.component.button.SubmitLabelButton
import net.dontdrinkandroot.wicket.bootstrap.component.form.RepeatingAjaxForm
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputEmail
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputPassword
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupInputText
import net.dontdrinkandroot.wicket.bootstrap.component.form.formgroup.FormGroupSelect
import net.dontdrinkandroot.wicket.bootstrap.css.Spacing
import net.dontdrinkandroot.wicket.kmodel.kModel
import net.dontdrinkandroot.wicket.kmodel.kModelValue
import net.dontdrinkandroot.wicket.kmodel.writableProperty
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.markup.repeater.RepeatingView
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model
import org.apache.wicket.model.StringResourceModel
import org.apache.wicket.model.util.ListModel
import org.apache.wicket.spring.injection.annot.SpringBean
import java.time.ZoneId
import java.util.*

open class UserEditForm(id: String, model: IModel<User>) : RepeatingAjaxForm<User>(id, model) {

    @SpringBean
    private lateinit var userService: UserService

    private val newPasswordModel: IModel<String> = Model()

    override fun populateFormGroups(formGroupView: RepeatingView) {
        val formGroupFirstName = FormGroupInputText(
            formGroupView.newChildId(),
            kModel.writableProperty(User::firstName),
            StringResourceModel("gitki.user.firstName")
        )
        formGroupFirstName.setRequired(true)
        formGroupFirstName.addAjaxValidation()
        formGroupView.add(formGroupFirstName)

        val formGroupLastName = FormGroupInputText(
            formGroupView.newChildId(),
            kModel.writableProperty(User::lastName),
            StringResourceModel("gitki.user.lastName")
        )
        formGroupLastName.setRequired(true)
        formGroupLastName.addAjaxValidation()
        formGroupView.add(formGroupLastName)

        val formGroupEmail = FormGroupInputEmail(
            formGroupView.newChildId(),
            kModel.writableProperty(User::email),
            Model.of("Email")
        )
        formGroupEmail.setRequired(true)
        formGroupEmail.addAjaxValidation()
        formGroupView.add(formGroupEmail)

        val formGroupLanguage = FormGroupSelect(
            formGroupView.newChildId(),
            kModel.writableProperty(User::language),
            StringResourceModel("gitki.user.language"),
            ListModel(Locale.getISOLanguages().toList()),
            IsoLanguageChoiceRenderer()
        )
        formGroupLanguage.setRequired(true)
        formGroupView.add(formGroupLanguage)

        val availableZoneIdsModel: IModel<List<String>> = ListModel(
            ZoneId.getAvailableZoneIds().toList()
                .sortedWith { obj: String, anotherString: String -> obj.compareTo(anotherString) }
        )
        val formGroupZoneId = FormGroupSelect(
            formGroupView.newChildId(),
            kModel.writableProperty(User::zoneId),
            StringResourceModel("gitki.timezone"),
            availableZoneIdsModel,
            ZoneIdChoiceRenderer()
        )
        formGroupZoneId.setRequired(true)
        formGroupView.add(formGroupZoneId)

        val formGroupRole = FormGroupSelect(
            formGroupView.newChildId(),
            kModel.writableProperty(User::role),
            StringResourceModel("gitki.user.role"),
            ListModel(Role.values().toList())
        )
        formGroupRole.setNullValid(false)
        formGroupRole.setRequired(true)
        formGroupRole.isVisible = getGitkiSession().hasRole(Role.ADMIN)
        formGroupView.add(formGroupRole)

        val formGroupPassword = FormGroupInputPassword(
            formGroupView.newChildId(),
            newPasswordModel,
            StringResourceModel("gitki.user.password")
        )
        formGroupPassword.setRequired(false)
        formGroupView.add(formGroupPassword)

        formGroupView.forEach { it.add(CssClassAppender(Spacing.MARGIN_BOTTOM_FULL)) }
    }

    override fun populateActions(buttonView: RepeatingView) {
        buttonView.add(SubmitLabelButton(buttonView.newChildId(), StringResourceModel("gitki.save")))
    }

    override fun onSubmit(target: AjaxRequestTarget?) {
        super.onSubmit(target)
        var user = this.kModelValue
        user = userService.save(user, newPasswordModel.getObject())

        /* Update user in session so changes are reflected e.g. for locale*/
        if (getCurrentUser()?.equals(user) == true) {
            getGitkiSession().signIn(user)
        }
        this.session.success(this.getString("gitki.user.saved"))
    }
}